Lib/test/test_posixpath.py

```diff
@@ -660,6 +660,20 @@ def test_realpath_resolve_first(self):
             safe_rmdir(ABSTFN + "/k")
             safe_rmdir(ABSTFN)
 
+    @os_helper.skip_unless_symlink
+    @skip_if_ABSTFN_contains_backslash
+    @unittest.skipIf(os.chmod not in os.supports_follow_symlinks, "Can't set symlink permissions")
+    def test_realpath_unreadable_symlink(self):
+        try:
+            os.symlink(ABSTFN+"1", ABSTFN)
+            os.chmod(ABSTFN, 0o000, follow_symlinks=False)
+            self.assertEqual(realpath(ABSTFN), ABSTFN)
```

Comment:
Please test also cases where non-readable symlink is followed by  other path components (including `..`), e.g. `badlink/foo`, `badlink/../foo`, `badlink/foo/..`, etc.

If the following `..` removes `badlink`, try to test something that involves a loop containing `badlink`, if this is possible. Your code does not set `seen[newpath] = None` before calling `readlink()`, I wonder what effect does it have.

Reply:
This is the behaviour we would be emulating:

```zsh
wannes@Stefans-iMac ~ % ln -s . src        
wannes@Stefans-iMac ~ % grealpath -e src/..
/Users
wannes@Stefans-iMac ~ % chmod -h 000 src
wannes@Stefans-iMac ~ % grealpath -e src/..
/Users/wannes
```

Is that not a problem, unlike the behaviour of `abspath()` compared to `realpath()`?

Reply:
```suggestion
            self.assertEqual(realpath(ABSTFN), ABSTFN)
            self.assertEqual(realpath(ABSTFN + '/foo'), ABSTFN + '/foo')
            self.assertEqual(realpath(ABSTFN + '/../foo'), dirname(ABSTFN) + '/foo')
            self.assertEqual(realpath(ABSTFN + '/foo/..'), ABSTFN)
```